### PR TITLE
[Core - Autoscaler] Upon autoscaler failure, propagate error message to all current and future drivers

### LINKED ISSUE #13538: Propagate monitor.py crash to driver logs
When the monitor crashes (e.g., unrecoverable autoscaling fault), we should propagate the error to drivers. This means
1. Write some error summary pointing to the logs to all active jobs.
2. Set a redis key so that future jobs will also see the error when connecting to the cluster.

cc @AmeerHajAli @DmitriGekhtman

### PR DESCRIPTION
**Pull request by clarkzinzow**
## Why are these changes needed?

Previously, an autoscaler failing would notify all existing drivers that were currently connected to the cluster, but all future drivers would never see such an error. This PR ensures that all future drivers also see that autoscaling error.

### TODO:

- [x] Add test.

## Related issue number

Closes #13538 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/worker.py [hunk 1]
```diff
@@ -1082,7 +1085,12 @@ def listen_error_messages_raylet(worker, task_error_queue, threads_stopped):
     worker.error_message_pubsub_client.psubscribe(error_pubsub_channel)
 
     try:
-        # Get the errors that occurred before the call to subscribe.
+        if _internal_kv_initialized():
```
Comment (rkooo567) [inline python/ray/worker.py:1088]: Btw, is this creating a duplicated connection or sharing the existing one?

Comment (ericl) [reply python/ray/worker.py:1088]: We shouldn't need to check initialized right?

Comment (clarkzinzow) [reply python/ray/worker.py:1088]: It's just a worker mode check, and all following calls should reuse the single Redis client on the worker.

**Comment by ericl** [comment]
LGTM; if not possible to add tests (perhaps in test_multi_node_2.py), can you try testing it out manually to verify?

**Review by wuisawesome — DISMISSED** [review]
Ah just realized, can we put the monitor initialization (and thus autoscaler initialization) in the try/catch so that we catch config issues?

**Comment by clarkzinzow** [comment]
@wuisawesome That's a great point! I'll do that.

**Comment by clarkzinzow** [comment]
@ericl I'm going to try to come up with an automated test, but I'll definitely test it manually at a minimum.

File: python/ray/monitor.py [hunk 2]
```diff
@@ -106,18 +108,19 @@ def __init__(self,
...
+
+    def _initialize_autoscaler(self):
+        if self.autoscaling_config:
             self.autoscaler = StandardAutoscaler(
-                autoscaling_config,
+                self.autoscaling_config,
                 self.load_metrics,
-                prefix_cluster_info=prefix_cluster_info,
+                prefix_cluster_info=self.prefix_cluster_info,
                 event_summarizer=self.event_summarizer)
```
Comment (clarkzinzow) [inline python/ray/monitor.py:123]: @wuisawesome I opted to separate out the autoscaler initialization from the constructor and move the failure handling into `Monitor.run()` in an effort to make it a bit cleaner.

File: python/ray/monitor.py [hunk 3]
```diff
@@ -235,13 +237,42 @@ def destroy_autoscaler_workers(self):
...
+
+    def _signal_handler(self, sig, frame):
+        return self._handle_failure(
+            f"Terminated with signal {sig}\n" +
+            "".join(traceback.format_stack(frame)))
+
     def run(self):
+        # Register signal handlers for autoscaler termination.
+        signal.signal(signal.SIGINT, self._signal_handler)
+        signal.signal(signal.SIGTERM, self._signal_handler)
```
Comment (clarkzinzow) [inline python/ray/monitor.py:267]: Signal handlers were easy to add and make it much easier to test the e2e failure handling (i.e. ensuring that the error makes it to the internal kv store and to the PubSub channel).

File: python/ray/worker.py [hunk 4]
```diff
@@ -851,7 +854,8 @@ def sigterm_handler(signum, frame):
 
 def custom_excepthook(type, value, tb):
     # If this is a driver, push the exception to GCS worker table.
-    if global_worker.mode == SCRIPT_MODE:
+    if global_worker.mode == SCRIPT_MODE and hasattr(
+            global_worker, "worker_id"):
```
Comment (clarkzinzow) [inline python/ray/worker.py:858]: I noticed that this was breaking in the autoscaler process.

**Review by ericl — APPROVED** [review]

**Comment by AmeerHajAli** [comment]
I just saw this PR (unfortunately was not assigned to review).
@clarkzinzow , can we add a follow up test that tests more complicated cases?
We want to test what happens if (1) ray driver attaches after a crash and (2) after a restart of autoscaler. We also wanna test  (3) if autoscaler does not send anything and ray driver attaches.

**Comment by clarkzinzow** [comment]
@AmeerHajAli Apologies for not assigning you! (1) is mostly covered by the existing test that ensures the autoscaler failure error is added to the internal kv store (which the driver checks upon connecting to the cluster). For (2) and (3), I agree that we should have a Python test for each of those cases, although I did test each of those cases manually:
2. Upon restarting, the autoscaler unsets the internal kv store error entry so that future drivers will not print out an autoscaler failure message. Note that this does _not_ log an "autoscaler recovered" message since I thought that was out-of-scope. It would be trivial to add by logging [here](https://github.com/ray-project/ray/blob/master/python/ray/monitor.py#L268-L269) with a `":recovery:"` token prefix and adding a check [here](https://github.com/ray-project/ray/blob/master/python/ray/worker.py#L961-L969) for that token, and logging out the recovery message if found.
3. If the autoscaler never sends anything, the prior driver behavior is preserved: the driver doesn't log out an error and proceeds as normal.

Is pushing an autoscaler recovery message to all drivers a requirement for you? If so, I can add that, and in that PR I can try to add some more Python tests covering (2) and (3).

**Comment by AmeerHajAli** [comment]
I think the tests are important to add regardless and it will be great if you can add the 2 tests.
I think sending the autoscaler recovered message is of lower priority (unless it is easier to do).
I think in general internally since we constantly restart the autoscaler container , it might be a useful feature.

**Comment by wuisawesome** [comment]
2 and 3 aren't really important to have in OSS since the autoscaler doesn't have a restart mechanism. this is an observability feature.

**Comment by AmeerHajAli** [comment]
@wuisawesome , these tests are important for when autoscaler is not in the user's container.

**Pull request merged**